### PR TITLE
[Merged by Bors] - feat: port data.rat.cast

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
- [x] depends on: #1159
- [x] depends on: #1251

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1159
* leanprover-community/mathlib4#1251
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1159~~
* ~~leanprover-community/mathlib4#1251~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Rat/Cast.lean [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+  (cast_def _).trans <| show (n / (1 : ℕ) : α) = n by rw [Nat.cast_one, div_one]
+#align rat.cast_coe_int Rat.cast_coe_int
+
+
+@[simp, norm_cast]
+theorem cast_coe_nat (n : ℕ) : ((n : ℚ) : α) = n := by
+  rw [← Int.cast_ofNat, ←ofInt_eq_cast, cast_coe_int, Int.cast_ofNat]
+#align rat.cast_coe_nat Rat.cast_coe_nat
+
+-- porting note: removed `norm_cast` attribute as statement had no cast
```
[the remark this answers is not in the document]
Comment (ChrisHughes24) [reply Mathlib/Data/Rat/Cast.lean:58]: It means that `HasRatCast.ratCast` was not marked with the `coe` attribute.

File: Mathlib/Data/Rat/Cast.lean [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+  ⟨RingHom.ext_rat⟩
+#align rat.subsingleton_ring_hom Rat.subsingleton_ring_hom
+
+namespace MulOpposite
+
+variable [DivisionRing α]
+
+-- Porting note: removed `norm_cast` attribute
+@[simp]
+theorem op_rat_cast (r : ℚ) : op (r : α) = (↑r : αᵐᵒᵖ) := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Rat/Cast.lean:548]:
```suggestion
theorem op_ratCast (r : ℚ) : op (r : α) = (↑r : αᵐᵒᵖ) := by
```

File: Mathlib/Data/Rat/Cast.lean [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+  rw [cast_def, div_eq_mul_inv, op_mul, op_inv, op_natCast, op_int_cast,
+    (Commute.cast_int_right _ r.num).eq, cast_def, div_eq_mul_inv]
+#align mul_opposite.op_rat_cast MulOpposite.op_rat_cast
+
+-- Porting note: removed `norm_cast` attribute
+@[simp]
+theorem unop_rat_cast (r : ℚ) : unop (r : αᵐᵒᵖ) = r := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Rat/Cast.lean:555]:
```suggestion
theorem unop_ratCast (r : ℚ) : unop (r : αᵐᵒᵖ) = r := by
```

File: Mathlib/Data/Rat/Cast.lean [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+theorem cast_hom_rat : castHom ℚ = RingHom.id ℚ :=
+  RingHom.ext cast_id
+#align rat.cast_hom_rat Rat.cast_hom_rat
+
+end Rat
+
+open Rat
+
+@[simp]
+theorem map_rat_cast [DivisionRing α] [DivisionRing β] [RingHomClass F α β] (f : F) (q : ℚ) :
```
Comment (ChrisHughes24) [inline Mathlib/Data/Rat/Cast.lean:486]:
```suggestion
theorem map_ratCast [DivisionRing α] [DivisionRing β] [RingHomClass F α β] (f : F) (q : ℚ) :
```

File: Mathlib/Data/Rat/Cast.lean [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+    f q = q := by rw [cast_def, map_div₀, map_int_cast, map_natCast, cast_def]
+#align map_rat_cast map_rat_cast
+
+@[simp]
+theorem eq_rat_cast {k} [DivisionRing k] [RingHomClass F ℚ k] (f : F) (r : ℚ) : f r = r := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Rat/Cast.lean:491]:
```suggestion
theorem eq_ratCast {k} [DivisionRing k] [RingHomClass F ℚ k] (f : F) (r : ℚ) : f r = r := by
```

File: Mathlib/Data/Rat/Cast.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (siddhartha-gadgil) [reply Mathlib/Data/Rat/Cast.lean]: I have added the `coe` attribute and restored the `norm_cast` attribute in all cases. As a test proofs that were previously `by norm_cast` work once more.

**Comment by ChrisHughes24** [comment]
bors merge

**Comment by Ruben-VandeVelde** [comment]
Thanks for the help!

**Pull request closed without merging**